src/test/java/com/emc/object/s3/S3JerseyClientTest.java

```diff
@@ -2971,4 +2973,182 @@ protected void assertAclEquals(AccessControlList acl1, AccessControlList acl2) {
+
+        // Object and associated multiple tags should get deleted
+        Assert.assertEquals(2, client.getObject(new GetObjectRequest(bucketName, key).withVersionId(versionId1), String.class).getObjectMetadata().getTaggingCount());
+        client.deleteObject(new DeleteObjectRequest(bucketName, key).withVersionId(versionId1));
+        try {
+            client.getObjectTagging(new GetObjectTaggingRequest(bucketName, key).withVersionId(versionId1));
+            Assert.fail("Fail was expected. Can NOT get tags from a deleted object");
+        } catch (S3Exception e) {
+            Assert.assertEquals(404, e.getHttpCode());
+            Assert.assertEquals("NoSuchKey", e.getErrorCode());
+        }
+
+    }
+
+    @Test
+    public void testCopyObjectWithTagging() {
+
+        // set up env
+        String bucketName = getTestBucket(), key1 = "test-object-tagging-src", key2 = "test-object-tagging-dest1", key3 = "test-object-tagging-dest2", content = "Hello Object Tagging!";
+
+        // should be able to copy the object and copied object should have the tags also
+        client.putObject(new PutObjectRequest(bucketName, key1, content)
+                .withObjectTagging(new ObjectTagging().withTagSet(Arrays.asList(new ObjectTag("k11", "v11"), new ObjectTag("k12", "v12")))));
+        client.copyObject(new CopyObjectRequest(bucketName, key1, bucketName, key2));
+        client.copyObject(new CopyObjectRequest(bucketName, key1, bucketName, key2));
+        Assert.assertEquals(2, client.getObjectTagging(new GetObjectTaggingRequest(bucketName, key2)).getTagSet().size());
+
+        // Versioned object should be copied and user should be able to get the same along with tags
+        client.setBucketVersioning(bucketName, new VersioningConfiguration().withStatus(VersioningConfiguration.Status.Enabled));
+        client.putObject(new PutObjectRequest(bucketName, key1, content)
+                .withObjectTagging(new ObjectTagging().withTagSet(Arrays.asList(new ObjectTag("k11", "v11"), new ObjectTag("k12", "v12")))));
+        String versionId = client.listVersions(bucketName, null).getVersions().get(0).getVersionId();
+        client.copyObject(new CopyObjectRequest(bucketName, key1, bucketName, key2).withSourceVersionId(versionId));
+        Assert.assertEquals(2, client.getObjectTagging(new GetObjectTaggingRequest(bucketName, key2).withVersionId(versionId)).getTagSet().size());
+
+    }
+
+    @Test
+    public void testMultipartUploadWithTagging() {
+
+        // set up env
+        String bucketName = getTestBucket(), key = "test-object-tagging";
+        int K = 1024;
+        IntStream ints = new Random().ints(3, K, 2 * K);
+        List<Integer> list = ints.boxed().collect(Collectors.toList());
+        int len = ints.sum(), len1 = list.get(0), len2 = list.get(1), len3 = list.get(2);
```

Comment:
not sure this is doing what you think - len2 and len3 may be smaller than len1... might be simpler to just have 3 different random byte arrays.

Reply:
Note that there is a `RandomInputStream` you can use here

Reply:
RandomInputStream introduced.